picard/profile.py

```diff
@@ -0,0 +1,125 @@
+
+# Imported to trigger inclusion of N_() in builtins
+from picard import i18n  # noqa: F401,E402 # pylint: disable=unused-import
+
+
+class UserProfileGroups():
+    """Provides information about the profile groups available for selecting in a user profile,
+    and the title and settings that apply to each profile group.
+    """
+    PROFILE_GROUPS = OrderedDict()  # Add groups in the order they should be displayed
+
+    PROFILE_GROUPS["metadata"] = {
+        "title": N_("Metadata"),
+        "settings": [
+            "va_name",
+            "nat_name",
+            "artist_locale",
+            "translate_artist_names",
+            "release_ars",
+            "track_ars",
+            "convert_punctuation",
+            "standardize_artists",
+            "standardize_instruments",
+            "guess_tracknumber_and_title",
+        ],
+    }
+
+    PROFILE_GROUPS["tags"] = {
+        "title": N_("Tags"),
+        "settings": [
+            "dont_write_tags",
+            "preserve_timestamps",
+            "clear_existing_tags",
+            "preserve_images",
+            "remove_id3_from_flac",
+            "remove_ape_from_mp3",
+            "preserved_tags",
+            "aac_save_ape",
+            "remove_ape_from_aac",
+            "ac3_save_ape",
+            "remove_ape_from_ac3",
+            "write_id3v1",
+            "write_id3v23",
+            "id3v2_encoding",
+            "id3v23_join_with",
+            "itunes_compatible_grouping",
+            "write_wave_riff_info",
+            "remove_wave_riff_info",
+            "wave_riff_info_encoding",
+        ],
+    }
+
+    PROFILE_GROUPS["coverart"] = {
+        "title": N_("Cover Art"),
+        "settings": [
+            "save_images_to_tags",
+            "embed_only_one_front_image",
+            "save_images_to_files",
+            "cover_image_filename",
+            "save_images_overwrite",
+            "save_only_one_front_image",
+            "image_type_as_filename",
+            "ca_providers",
+        ],
+    }
+
+    PROFILE_GROUPS["filenaming"] = {
+        "title": N_("File Naming"),
+        "settings": [
+            "windows_compatibility",
+            "ascii_filenames",
+            "rename_files",
+            "move_files",
+            "move_files_to",
+            "move_additional_files",
+            "move_additional_files_pattern",
+            "delete_empty_dirs",
+        ],
+    }
+
+    PROFILE_GROUPS["scripting"] = {
+        "title": N_("Scripting"),
+        "settings": [
+            "enable_tagger_scripts",
+            "list_of_scripts",
+            "last_selected_script_position",
+        ],
+    }
+
+    @classmethod
+    def get_profile_settings_list(cls):
+        """Iterable of all settings names in all profile groups.
+
+        Yields:
+            str: Setting name
+        """
+        for key, profile_group in cls.PROFILE_GROUPS.items():
+            for setting in profile_group["settings"]:
```

Comment:
What if 2 profile groups contain the same setting?

Reply:
They actually shouldn't, but even if they do the setting would appear twice in the list.  The problem would come if the user had both groups selected in the same profile and then removed one of the groups, where it might be possible that the setting would be removed from that profile.  When the time comes to do the setting removal, the code should take that into account.  Good catch!

Reply:
This was the whole point of what I called "stack", the idea behind this is to let user (at some point) defines whichever group of settings he wants, and applies first one met.

If we have group 1 (G1) with setting (S1), and group 2 with same setting but different value (S2), along default setting (SD), and all groups are enabled, we iterate from top to bottom (bottom being default):

Enabled groups order:
```
G2 -> G1 -> D
```

```
S2? -> no -> S1? -> no -> use SD
       | yes         | yes
       |- use S2     |-use S1
```

Reply:
I think we're using the same terms for different things.  To clarify my use:

**profile** is an item in the **profile stack**.  Each profile contains zero or more **settings** which are used to override the base settings if the profile is active.  When retrieving a value from `config.setting`, the system looks at each profile in the stack from top to bottom (i.e. first to last in the `config.setting["user_profiles"]` list).  If the profile is "active" and contains a value (which is not `None`) for the requested setting, then that value is returned and the profile review process terminates. If there is no active profile that provides the requested setting, then the value from `config.setting` is returned.

When saving a setting, the system again looks at each profile in the stack from top to bottom (i.e. first to last in the `config.setting["user_profiles"]` list).  If an "active" profile is found, and it contains an item in its "settings" dictionary for the setting being saved, then that profile is updated with the new setting and the process review process terminates, with the updated profile saved to `config.setting["user_profiles"]`.  If there is no active profile that includes the setting in its "settings" dictionary, then the value in `config.setting` is updated.

I think that we share the same view as to how the profile stack and setting assignment and retrieval would work.

**setting groups** are used to logically separate the settings into lists of related settings, and are primarily intended to allow the user to add/remove settings to/from a profile as a group rather than specifying each individual setting.  For example, when the user is creating or editing a profile, they could be presented with a tree view of settings available for selection with the setting groups as the branches and the settings within each setting group as the leaves.  Selection/deselection could be done with checkboxes at the branch and leaf level in the tree view.  When a setting is selected for a profile, that setting would be added to the profile's "settings" dictionary with an initial value of `None`.  When a setting is deselected from a profile, that setting would be removed from the profile's "settings" dictionary.

Depending on how we decide to proceed, potentially only the branch level would be displayed for selection.  In this case, when a setting group is selected for a profile, each of the settings associated with that settings group would be added to the profile's "settings" dictionary with an initial value of `None`.  When a settings group is deselected from a profile, each of the settings associated with that settings group would be removed from the profile's "settings" dictionary.

Reply:
> setting groups

That's what you named "profile groups" in the code, right? Perhaps it would be clearer if named "setting groups", according to the description above those don't have much to do with profiles (apart the fact they are used by them).

Reply:
That makes a lot of sense, and should help clear up some of the confusion.  I'll make the change.